### ISSUE TITLE
How to prevent loss of case in company-ispell with grouped backends

### ISSUE THREAD

**Issue by mohkale** [company-mode/company-mode#1096]
I'm using company-ispell to suggest valid words when entering text in org-mode and similair buffers. The problem I'm facing at the moment is when I try to enter a capitalised word `company-ispell` suggests the lowercase variant and if I accept it then it overwrites my captialised version with the lowercase one.

For example

```
* Hello Worl<point>
```

At point I start company and it suggests several words including `world` as a completion candidate. I select it and end up with:

```
* Hello world
```

Where I actually wanted:

```
* Hello World
```

There is an option `company-dabbrev-ignore-case` for dabbrev which I believe tries to solve a similair issue. Is there an equivalent for company-ispell?

## Version Information
- `uname -a`: `Linux mk-desktop 5.11.13-arch1-1 #1 SMP PREEMPT Sat, 10 Apr 2021 20:47:14 +0000 x86_64 GNU/Linux`
- emacs-version: `GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.28, cairo version 1.17.4) of 2021-04-09`
- company-version: `9bc678a`

**Comment by dgutov** [comment]
Hi!

`company-ispell` already returns `keep-prefix` to `ignore-case`.

Shouldn't this just work? If I type `Hello W` and follow with `M-x company-ispell`, all the suggestions are capitalized.

Perhaps some other backend is in play?

**Comment by mohkale** [comment]
@dgutov 

You're correct, with `emacs -q` and company-backends set to just `'(company-ispell)` it works. However here's my value for company-backends with which it doesn't seem to work.

```lisp
((:separate company-capf company-yasnippet company-abbrev company-ispell)
 (company-capf company-yasnippet)
 company-capf)
```

**Comment by dgutov** [comment]
All right, that's due to 288a7617 (see the linked reports that it closes).

Perhaps it's time to eliminate the `keep-prefix` special case, and instead do the case conversion inside each respective backend, when configured to do so. I'm guessing we didn't do that originally because of the associated performance overhead, but the edge cases are a pain.

With this, we could also fix the completion for the ALLCAPS case as well (which had never been reported).

Would you like to try implementing it?

Basically, backends like `company-dabbrev` and `company-ispell` will now postprocess their `candidates` before returning them to contain the expected letter case for the whole word. Not sure if there are examples to copy the logic from: `replace-match` does that when FIXEDCASE is nil, but the implementation is in C.

**Comment by mohkale** [comment]
@dgutov 

> Would you like to try implementing it?

I'm not sure, this seems to be more related to company internals which I don't have a good enough understanding of to approach atm.

For now I'll just remove company-ispell from my backends to avoid completion messing up my writing?

**Comment by dgutov** [comment]
> this seems to be more related to company internals which I don't have a good enough understanding of to approach atm

It will be implemented in the backend code: `company-dabbrev` and `company-ispell`. The company internals will just see a removal of an existing feature. Well, and maybe we could extract a helper, to share code between the backends.

No pressure, though.

**Comment by c02y** [comment]
`company-dabbrev" has `company-dabbrev-downcase`, maybe variable like `company-ispell-downcase` should be provided for `company-ispell`, it is quite useful when writing org-mode files.

**Comment by dgutov** [comment]
`company-dabbrev-downcase` decided whether all completions are converted to lower case.

Is there a point in downcasing `company-ispell`'s suggestions? Most of them are already lower case, and those that aren't, are all proper nouns.

**Comment by yugaego** [comment]
Related Q&A #1226.
